File: go/passgen/passgen.go
package main

import (
	"crypto/sha512"
	"encoding/hex"
	"flag"
	"fmt"
	"math/rand"
	"os"
	"slices"
	"strconv"
	"strings"
	"time"
)

func replaceAtIndex(str string, pos int, char byte) string {
	return str[:pos] + string(char) + str[pos+1:]
}

func toggleCase(char byte) byte {
	if strings.ToUpper(string(char)) == string(char) {
		return byte(strings.ToLower(string(char))[0])
	}

	return byte(strings.ToUpper(string(char))[0])
}

func generatePassword(length int, separator string) string {
	password := ""

	for i := 1; i <= length; i++ {
		if separator != "" && length > 6 && i%6 == 0 && i != length {
			password += separator
			continue
		}

		randInt := rand.Intn(length)
		now := time.Now().Unix() * int64(randInt)

		hash := sha512.New()
		hashText := hex.EncodeToString(hash.Sum([]byte(strconv.FormatInt(now, 10))))

		randIndex := rand.Intn(len(hashText))
		password += string(hashText[randIndex])
	}

	for range password {
		randIndex := rand.Intn(max(len(password)-1, 1))
		password = replaceAtIndex(password, randIndex, toggleCase(password[randIndex]))
	}

	return password
}

func findArg(argv []string) (string, []string) {
	for i := range argv {
		if (i == 0 && argv[i][0] != '-') || (i-1 >= 0 && argv[i-1][0] != '-') {
			return argv[i], slices.Concat(argv[:i], argv[i+1:])
		}
	}

	return "", argv
}

func main() {
	var separator string
	flag.StringVar(&separator, "s", "-", "The separator for blocks in the password.")

	arg, remainingArgs := findArg(os.Args[1:])

	flag.CommandLine.Parse(remainingArgs)

	if arg == "" {
		fmt.Println("USAGE: passgen <length> [-s <s>]")
		os.Exit(1)
	}

	length, err := strconv.Atoi(arg)
	if err != nil {
		fmt.Println("\033[31mError parsing arguments: not an integer:\033[0m", arg)
		os.Exit(1)
	}

	password := generatePassword(length, separator)
	fmt.Println(password)
}

File: swift/passgen/passgen.swift
import CryptoKit
import Foundation

extension String {
    subscript(_ i: Int) -> Character {
        get {
            return Array(self)[i]
        }
        set(newValue) {
            var arr = Array(self)
            arr[i] = newValue
            self = String(arr)
        }
    }

    mutating func replaceAtIndex(_ i: Int, _ char: Character) {
        self[i] = char
    }
}
extension Character {
    func toggleCase() -> Self {
        if isLowercase {
            return Character(uppercased())
        }
        return Character(lowercased())
    }
}

func generatePassword(_ length: Int, _ separator: String) -> String {
    var password = ""

    for i in 0..<length {
        if separator != "" && length > 6 && i % 6 == 0 && i != length {
            password += separator
            continue
        }

        let randInt = Int.random(in: 0..<length)
        let now = Int(Date().timeIntervalSince1970) * randInt

        let hash = SHA3_512.hash(data: "\(now)".data(using: .unicode)!).description
        let hashText = String(hash.suffix(128))

        let randIndex = Int.random(in: 0..<hashText.count)
        password += String(hashText[randIndex])
    }

    for _ in password {
        let randIndex = Int.random(in: 0..<max(password.count - 1, 1))
        password.replaceAtIndex(randIndex, password[randIndex].toggleCase())
    }

    return password
}

struct Args {
    var help = false
    var separator = "-"
    var length = 0

    mutating func parse() {
        var remainingArgs: [String] = []
        let argv = CommandLine.arguments
        var skip = false

        for i in 1..<Int(CommandLine.argc) {
            if skip {
                skip = false
                continue
            }
            if argv[i] == "--help" || argv[i] == "-h" {
                help = true
                continue
            }
            if argv[i] == "-s" {
                separator = argv[i + 1]
                skip = true
                continue
            }
            if !argv[i].starts(with: "-") {
                remainingArgs.append(argv[i])
            }
        }

        if remainingArgs.count > 0 {
            length = Int(remainingArgs[0])!
        }
    }
}

var args = Args()
args.parse()

if args.help {
    print(
        """
        USAGE: \(CommandLine.arguments[0]) <length> [-s <separator>]

        ARGUMENTS:
          <length>      The length of the password to generate.

        OPTIONS:
          -s            The separator for blocks in the password (default: -)
          -h, --help    Show help information.
        """)
    exit(0)
}

let password = generatePassword(args.length, args.separator)
print(password)
